substantial/types.py

```diff
@@ -118,10 +140,37 @@ async def exec(self, ctx, save_id, counter) -> Any:
             ctx.source(events.Event(save=save))
             return ret
         except Exception as e:
+            if self.compensate_fn:
+                compesation_attemps = 0
+                while compesation_attemps < self.max_compensation_attempts:
+                    try:
+                        compensation_result = self.compensate_fn()
+                        ctx.source(
+                            events.Event(
```

Comment:
Maybe allow async functions, wdyt?

Reply:
Also design-wise, what do you think should happen when a save step is compensated, should we progress further or stop right there?

Reply:
I'm not sure to understand what u meaning!

Reply:
I am seeing here that the execution is still proceeding and not throw the latest error but instead return the compensate fn value, I was wondering if this was an intentional part of your design.

Reply:
What I'm doing is, run the function compensate if there is an exception in lambda_fn,   and just get the result ? 
and if compensation is raise exception.

Reply:
Shouldn't "compensate" reflect the idea of a [rollback](https://en.wikipedia.org/wiki/Rollback_(data_management))?